### PR TITLE
Symbolic debugger

### PR DESCRIPTION
**Pull request by shomilj**
Add (the beta version of) the symbolic debugger and improved logging to userlib. This adds a symbols table that keeps track of where all of the cryptographic strings involved in a project come from for ease of understanding and debugging. Based on testing staff and past student solutions, we believe it works for most use cases. It's a known issue that certain designs will cause problems for this (when different inputs generate the same cryptographic strings), but correct designs should never result in this behavior anyway. We intend to field feedback as the project goes to implement patches as needed.

### CODE REVIEW

File: userlib.go [hunk 1]
```diff
@@ -411,42 +588,167 @@ func symEnc(key []byte, iv []byte, plaintext []byte) []byte {
...
+	}
+	data_bytes, err := json.Marshal(truncated)
+	if err != nil {
+		panic("symbolic debugger failed in truncateStr with error: " + err.Error())
+	}
+	return fmt.Sprintf("%s", data_bytes)
+}
+
+// Converts byte slice data to valid symbolic map keys.
+func keyFromBytes(data []byte) (truncated string) {
```
Comment (conorgil) [inline userlib.go:706]: Suggestion to improve readability: rename this function to distinguish from an encryption key.
```
// Converts byte slice data to a fixed size value that is suitable to use as the
//storage key in a map.

func mapKeyFromBytes(data []byte) (truncated string) {
	return fmt.Sprintf("%x", sha512.Sum512(data))
}
```

**Review by conorgil — COMMENTED** [review]
Bunch of suggestions on readability improvements. Overall, looks like a very cool addition to make debugging easier! Kudos 🎉

**Comment by shomilj** [comment]
Incorporated most suggestions. Won't have time to write tests by tomorrow, but feel free to do so if you'd like. I think the current ones may be a bit broken.

File: userlib.go [hunk 2]
```diff
@@ -233,6 +348,24 @@ type PKEDecKey = PrivateKeyType
 type DSSignKey = PrivateKeyType
 type DSVerifyKey = PublicKeyType
 
+func recordKeys(publicKey rsa.PublicKey, privateKey rsa.PrivateKey,
+	publicKeyStruct interface{}, privateKeyStruct interface{},
+	publicKeyFormat string, privateKeyFormat string) {
+
+	publicKeyId := truncateBytes(x509.MarshalPKCS1PublicKey(&publicKey))
```
[the remark this answers is not in the document]
Comment (conorgil) [reply userlib.go]: I took the liberty of making this change. Please shout at me in Slack if you are offended.

File: userlib.go [hunk 3]
```diff
@@ -208,12 +273,62 @@ var Argon2Key = argon2Key
...
+// UUIDFromBytes creates a new UUID from a byte slice.
+// Returns an error if the slice does not have a length of 16.
+// The bytes are copied from the slice.
+func uuidFromBytes(b []byte) (result UUID, err error) {
+	if len(b) < 16 {
+		return uuid.New(), errors.New("UUIDFromBytes expects an input greater than or equal to 16 characters")
+	}
+	result, err = uuid.FromBytes(b[:16])
+	if err != nil {
+		return uuid.New(), err
```
[the remark this answers is not in the document]
Comment (conorgil) [reply userlib.go:322]: I re-read the uuid docs just now and realized that my understanding of `uuid.NullUUID` is incorrect. The code is best as-written.

**Review by conorgil — APPROVED** [review]

**Pull request merged**